src/JetMETAnalyzer.cc

```diff
@@ -0,0 +1,517 @@
+  betaStar.clear();
+  betaClassic.clear();
+  betaStarClassic.clear();
+  dZ.clear();
+  DRweighted.clear();
+  fRing0.clear();
+  fRing1.clear();
+  fRing2.clear();
+  fRing3.clear();
+  fRing4.clear();
+  fRing5.clear();
+  fRing6.clear();
+  fRing7.clear();
+  fRing8.clear();
+  nCh.clear();
+  nNeutrals.clear();
+  ptD.clear();
+  isMatched.clear();
+
+  size_t nJet = (nJetMax_ == 0) ? jets->size() : std::min(nJetMax_, (unsigned int) jets->size());
+  for (size_t iJet = 0; iJet < nJet; iJet++) {
+
+     pat::Jet const & jet = jets->at(iJet);
+     if (jet.pt() < 5)
+         continue;
+
+     const reco::GenJet* ref = jet.genJet();
+
+     if (ref) {
+       refdrjt.push_back( reco::deltaR(jet.eta(),jet.phi(),ref->eta(),ref->phi()) );
+       isMatched.push_back(true);
+     }
+     else {
+       refdrjt.push_back(0);
+       isMatched.push_back(false);
+     }
+
+     refrank.push_back( nref );
+     refpdgid_algorithmicDef.push_back( 0 );
+     refpdgid_physicsDef.push_back( 0 );
+     if(ref) { 
+        refpdgid.push_back( ref->pdgId() );
+        refe.push_back( ref->energy() );
+        refpt.push_back( ref->pt() );
+        refeta.push_back( ref->eta() );
+        refphi.push_back( ref->phi() );
+        refm.push_back( ref->mass() );
+        refy.push_back( ref->rapidity() );
+        refarea.push_back( ref->jetArea() );
+     }
+     else {
+        refpdgid.push_back( 0. );
+        refe.push_back( 0. );
+        refpt.push_back( 0. );
+        refeta.push_back( 0. );
+        refphi.push_back( 0. );
+        refm.push_back( 0. );
+        refy.push_back( 0. );
+        refarea.push_back( 0. );
+     }
+
+     jte.push_back( jet.energy() );
+     jtpt.push_back( jet.pt() );
+     jteta.push_back( jet.eta() );
+     jtphi.push_back( jet.phi() );
+     jtm.push_back( jet.mass() );
+     jty.push_back( jet.rapidity() );
+     jtarea.push_back( jet.jetArea() );
+     jtjec.push_back( jet.jecFactor(0) );
+
+     int n_pf = jet.numberOfDaughters();
+     float phiJet = jet.phi();
+     float etaJet = jet.eta();
+
+     int nCh_tmp(0), nNeutrals_tmp(0);
+     float sumTkPt(0.0);
+     float beta_tmp(0.0), betaStar_tmp(0.0), betaStarClassic_tmp(0.0), betaClassic_tmp(0.0);
+     float pTMax(0.0),dZ2(-999);
+     float sumW(0.0),sumW2(0.0),sumWdR2(0.0);
+     float sum_ring0(0.0),sum_ring1(0.0),sum_ring2(0.0),sum_ring3(0.0),sum_ring4(0.0),sum_ring5(0.0),sum_ring6(0.0),sum_ring7(0.0),sum_ring8(0.0);
+     float ptD_tmp(-1.0),DR_weighted_tmp(0.0);
+
+     for(int j=0;j<n_pf;j++) {
+         auto part = jet.daughterPtr(j);
+         if (!( part.isAvailable() && part.isNonnull()) ){
+             continue;	
+         }
+
+         if (fabs(part->charge()) > 0) {
+             nCh_tmp++;
+         }
+         else if(fabs(part->charge())==0){
+             nNeutrals_tmp++;
+         }
+
+         float deta = part->eta() - etaJet;
+         float dphi = 2*atan(tan((part->phi()-phiJet)/2));
+         float dR = sqrt(deta*deta + dphi*dphi);
```

Comment:
There are standard packages to do the deltaR calculation ("DataFormats/Math/interface/deltaR.h" --> reco::deltaR). Should we be using those? Saves a line or two.